python/revng/model/_util.py

```diff
@@ -0,0 +1,52 @@
+from enum import Enum
+
+
+def enum_value_to_index(enum_value: Enum):
+    """Converts an enum value to its index"""
+    return list(enum_value.__class__.__members__).index(enum_value.value)
+
+
+def autoassign_primitive_id(base_class: type):
+    """Monkeypatches the __init__ method so that the ID is automatically assigned.
+    Meant for use with primitive types.
+    """
+    assert isinstance(base_class, type)
+
+    original_init = base_class.__init__
+
+    def init_with_computed_id(self, *args, **kwargs):
+        if "ID" not in kwargs:
+            PrimitiveKind = kwargs.get("PrimitiveKind")
+            if PrimitiveKind is None:
+                raise ValueError("Must provide PrimitiveKind as kwarg to PrimitiveType constructor")
+
+            Size = kwargs.get("Size")
+            if Size is None:
+                raise ValueError("Must provide Size as kwarg to PrimitiveType constructor")
+
+            primitive_kind_value = enum_value_to_index(PrimitiveKind)
+            kwargs["ID"] = primitive_kind_value << 8 | Size
+        original_init(self, *args, **kwargs)
+
+    base_class.__init__ = init_with_computed_id
```

Comment:
I'd really like to get rid entirely of monkey patching. Can we export a `PrimitiveType` that inherits the generated `PrimitiveType` but overrides `__init__` in a simple and explicit way?

Reply:
I didn't do it that way because I thought that require code duplication when we'll have multiple versions of the model, but we can do something like this:

`common.py`
```python
def create_primitive_type_class(base_class):
    class PrimitiveType(base_class):
        def __init__(...):
            ...
    return PrimitiveType
```
    

`v1/__init__.py`
```python
from common import create_primitive_type_class
from _generated import PrimitiveType as GeneratedPrimitiveType

# Create a derived PrimitiveType
PrimitiveType = create_primitive_type_class(GeneratedPrimitiveType)
```

`v2/__init__.py` would be similar

what do you think?

Reply:
Mmmh, what about
```python
class PrimitiveType(GeneratedPrimitiveType, MyOtherTypeOverridingInit):
  pass
```
? Would this work?
I feel like we should strive for avoiding having functions create new types or monkeypatching. Autocompleters will become very unhappy.